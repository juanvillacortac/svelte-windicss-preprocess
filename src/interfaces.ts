export interface Node {
  start: number,
  end: number,
  type: string,
  value?: Text | Expression | ( Text | Expression)[],
}

export interface Attribute extends Node {
  type: 'Attribute',
  name: string,
  value: Text | Expression | ( Text | Expression)[],
}

export interface Directive extends Node {
  type: 'Directive',
  name: string,
  value: Text | Expression | ( Text | Expression)[],
}

export interface Expression extends Node {
  type: 'Expression',
  data: string
}

export interface Text extends Node {
  type: 'Text',
  data: string
}

export interface Tag {
  start: number,
  end: number,
  name: string,
  value: (Attribute | Directive)[]
}