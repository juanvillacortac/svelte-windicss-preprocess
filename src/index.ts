import type { IconifyJSON } from '@iconify/types'
// import { useConfig, useDebugger } from '@nbhr/utils'
import type { UnoGenerator } from '@unocss/core'
import { createGenerator } from '@unocss/core'
import UnocssIcons from '@unocss/preset-icons'
import fg from 'fast-glob'
import { readFileSync } from 'fs'
import { parse, preprocess } from 'svelte/compiler'
import type {
  PreprocessorGroup,
  Processed,
} from 'svelte/types/compiler/preprocess'
import { Processor } from 'windicss/lib'
// import type { FullConfig } from 'windicss/types/interfaces'
import { CSSParser } from 'windicss/utils/parser'
import { StyleSheet } from 'windicss/utils/style'
import { globalStyleSheet, Magician, SetObject } from './utils'

// const DEV = false
// let windiConfig: FullConfig
// let configMTime: number
// let OPTIONS: BaseConfig

// function loadConfig(path: string): Promise<void> {
//   useDebugger.createLog('Trying to load windi configuration from ' + path)
//   return useConfig.load<FullConfig>(path).then(config => {
//     // write current unix timestamp to configMTime
//     configMTime = Date.now()
//     if (config.preflight === false) OPTIONS.preflights = false
//     if (config.safelist && typeof config.safelist == 'string') {
//       OPTIONS.safeList = config.safelist
//     } else if (config.safelist) {
//       const tmpSafelist = config.safelist as (string | string[])[]
//       OPTIONS.safeList = [...new Set(tmpSafelist.flat(Infinity))].join(' ')
//     }
//     console.log(config)
//     console.log(JSON.stringify(config.theme))
//     PROCESSOR.loadConfig(config)
//     useDebugger.createLog('Configuration loaded successfully')
//     windiConfig = config
//   })
// }

let entryFileName = ''

interface generatorObject {
  data: SetObject
  updatedAt: number
  writtenAt?: number
}

const raw = new Map<string, generatorObject>([
  [
    '__GLOBAL',
    {
      data: {
        inlineClasses: new Set(),
        inlineDirectives: new Set(),
        inlineExpressions: new Set(),
        inlineIcons: new Set(),
        inlineAttributify: new Map(),
      },
      updatedAt: Date.now(),
    },
  ],
])

function agent(): PreprocessorGroup {
  let result: SetObject
  return {
    async markup({ content, filename }): Promise<Processed> {
      let worker = new Magician(content, filename, configuration)
      worker = worker.prepare()
      // worker = worker.setInject()
      worker = worker.extract()
      result = worker.getSets()
      raw.set(filename, {
        data: result,
        updatedAt: Date.now(),
      })

      return {
        code: content,
      }
    },
  }
}

async function generateCSS(
  key: string,
  attributes: Record<string, string | boolean>
) {
  const t = raw.get(key)

  // MARK: WINDI DEFAULT
  //   const FILESHEET = CSS_STYLESHEETS.get(filename || '') || undefined
  //   if (FILESHEET && attributes['windi:global']) {
  //     INLINE_STYLE = globalStyleSheet(FILESHEET['code']).build()
  //   } else if (FILESHEET) {
  //     INLINE_STYLE = FILESHEET['code'].build()
  //   }
  const defaultStyleSheet = generatorWindi.interpret(
    Array.from(
      new Set([
        ...(t?.data.inlineClasses || new Set()),
        ...(t?.data.inlineDirectives || new Set()),
        ...(t?.data.inlineExpressions || new Set()),
      ])
    ).join(' ')
  ).styleSheet
  let defaultStyles = ''
  if (attributes['windi:global'] || attributes['windi-inline-global']) {
    defaultStyles = globalStyleSheet(defaultStyleSheet).build()
  } else {
    defaultStyles = defaultStyleSheet.build()
  }

  // MARK: ATTRIBUTIFY
  const nObj: Record<string, string[]> = {}
  t?.data.inlineAttributify.forEach((v: Set<string>, k: string) => {
    nObj[k] = Array.from(v)
  })

  const attributifyStyleSheet = generatorWindi.attributify(nObj).styleSheet
  let attributifyStyles = ''
  if (attributes['windi:global'] || attributes['windi-inline-global']) {
    attributifyStyles = globalStyleSheet(attributifyStyleSheet).build()
  } else {
    attributifyStyles = attributifyStyleSheet.build()
  }

  // MARK: ICONS (experimental)
  let iconStyles = ''
  if (generatorUno && configuration.experimental?.icons != undefined) {
    iconStyles = await generatorUno
      .generate(t?.data.inlineIcons || '')
      .then(resolve => resolve.css)
  }
  return { defaultStyles, attributifyStyles, iconStyles }
}

function main(): PreprocessorGroup {
  if (configuration.experimental?.scan !== undefined) {
    const files = fg.sync(['src/**/*.svelte'], {})
    for (const file of files) {
      const content = readFileSync(file).toString()
      const filename = file
      const ast = parse(content, { filename })
      const hasGlobalInline = ast.css.attributes.some(
        el => el.name == 'windi-inline-global'
      )
      console.log(filename, hasGlobalInline)
      let worker = new Magician(content, filename, configuration)
      worker = worker.prepare()
      worker = worker.extract()
      const result = worker.getSets()
      if (hasGlobalInline) {
        const global = raw.get('__GLOBAL')!.data
        raw.set('__GLOBAL', {
          data: {
            inlineClasses: new Set([
              ...global.inlineClasses,
              ...result.inlineClasses,
            ]),
            inlineDirectives: new Set([
              ...global.inlineDirectives,
              ...result.inlineDirectives,
            ]),
            inlineExpressions: new Set([
              ...global.inlineExpressions,
              ...result.inlineExpressions,
            ]),
            inlineIcons: new Set([
              ...global.inlineIcons,
              ...result.inlineIcons,
            ]),
            inlineAttributify: new Map([
              ...global.inlineAttributify,
              ...result.inlineAttributify,
            ]),
          },
          updatedAt: Date.now(),
        })
        raw.set(filename, {
          data: {
            inlineClasses: new Set(),
            inlineDirectives: new Set(),
            inlineExpressions: new Set(),
            inlineIcons: new Set(),
            inlineAttributify: new Map(),
          },
          updatedAt: Date.now(),
        })
      } else {
        raw.set(filename, {
          data: result,
          updatedAt: Date.now(),
        })
      }
    }
  }

  return {
    async style({ content, attributes, filename }): Promise<Processed> {
      if (!filename) return { code: content }

      //   // MARK: PREFLIGHTS
      //   if (OPTIONS.preflights === true && attributes['windi:preflights:global']) {
      //     const PREFLIGHTS = PROCESSOR.preflight()
      //     PREFLIGHTS_STYLE = globalStyleSheet(PREFLIGHTS).build()
      //   } else if (OPTIONS.preflights === true && attributes['windi:preflights']) {
      //     const PREFLIGHTS = PROCESSOR.preflight()
      //     PREFLIGHTS_STYLE = PREFLIGHTS.build()
      //   }
      let preflightStyleSheet = new StyleSheet()
      if (attributes['windi-preflights-global']) {
        preflightStyleSheet = globalStyleSheet(generatorWindi.preflight())
      }
      const preflightStyles = preflightStyleSheet.build()

      //   // MARK: SAFELIST
      //   if (OPTIONS.safeList && attributes['windi:safelist:global']) {
      //     const SAFELIST = PROCESSOR.interpret(OPTIONS.safeList).styleSheet
      //     SAFELIST_STYLE = globalStyleSheet(SAFELIST).build()
      //     if (OPTIONS.experimental && OPTIONS.experimental.icons != undefined) {
      //       let UNO_SAFELIST_STYLE = ''
      //       const { css } = await UNO.generate(OPTIONS.safeList)
      //       const UNO_SAFELIST_STYLESHEET = new CSSParser(css).parse()
      //       UNO_SAFELIST_STYLE = globalStyleSheet(UNO_SAFELIST_STYLESHEET).build()
      //       SAFELIST_STYLE += UNO_SAFELIST_STYLE
      //     }
      //   } else if (OPTIONS.safeList && attributes['windi:safelist']) {
      //     const SAFELIST = PROCESSOR.interpret(OPTIONS.safeList).styleSheet
      //     SAFELIST_STYLE = SAFELIST.build()
      //     if (OPTIONS.experimental && OPTIONS.experimental.icons != undefined) {
      //       let UNO_SAFELIST_STYLE = ''
      //       const { css } = await UNO.generate(OPTIONS.safeList)
      //       UNO_SAFELIST_STYLE = css
      //       SAFELIST_STYLE += UNO_SAFELIST_STYLE
      //     }
      //   }
      let safelistStyleSheet = new StyleSheet()
      if (attributes['windi-safelist-global']) {
        safelistStyleSheet = globalStyleSheet(
          generatorWindi.interpret(configuration.safeList || '').styleSheet
        )
      }
      const safelistStyles = safelistStyleSheet.build()

      let defaultStyles, attributifyStyles, iconStyles
      if (entryFileName.length > 0) {
        const result = await generateCSS(filename, attributes)
        defaultStyles = result.defaultStyles
        attributifyStyles = result.attributifyStyles
        iconStyles = result.iconStyles
      } else {
        entryFileName = filename
        const resultGlobal = await generateCSS('__GLOBAL', {
          'windi-inline-global': true,
        })
        defaultStyles = resultGlobal.defaultStyles
        attributifyStyles = resultGlobal.attributifyStyles
        iconStyles = resultGlobal.iconStyles
        const result = await generateCSS(filename, attributes)
        defaultStyles += result.defaultStyles
        attributifyStyles += result.attributifyStyles
        iconStyles += result.iconStyles
      }

      // MARK: CUSTOM CSS + WINDI @apply
      //   let CSS: StyleSheet
      //   CSS_SOURCE = content
      //   if (CSS_SOURCE && attributes['global']) {
      //     CSS = new CSSParser(CSS_SOURCE, PROCESSOR).parse()
      //     CSS_STYLE = globalStyleSheet(CSS).build()
      //   } else if (CSS_SOURCE) {
      //     const tmpCSS = CSS_SOURCE
      //     const rules = [...(tmpCSS.matchAll(/(?<selector>[^}]*){(?<css>[^}]*)}/gim) || [])]
      //     rules.forEach(rule => {
      //       if (rule.groups && rule.groups.selector.includes(':global')) {
      //         const globalCSS = new CSSParser(rule[0], PROCESSOR).parse()
      //         const buildGlobalCSS = globalStyleSheet(globalCSS).build()
      //         if (buildGlobalCSS.length > 0) CSS_STYLE += buildGlobalCSS + '\n'
      //       } else {
      //         CSS = new CSSParser(rule[0], PROCESSOR).parse()
      //         const buildLocalCSS = CSS.build()
      //         if (buildLocalCSS.length > 0) CSS_STYLE += buildLocalCSS + '\n'
      //       }
      //     })
      //   }
      const cssStyleSheet =
        new CSSParser(content, generatorWindi).parse() || new StyleSheet()

      let cssStyles = ''
      if (attributes['global']) {
        cssStyles = globalStyleSheet(cssStyleSheet).build()
      } else {
        cssStyles = cssStyleSheet.build()
      }

      //   // MARK: COMBINE
      //   let newStyleCode = '\n'
      //   if (PREFLIGHTS_STYLE.length > 0) newStyleCode += PREFLIGHTS_STYLE + '\n'
      //   if (SAFELIST_STYLE.length > 0) newStyleCode += SAFELIST_STYLE + '\n'
      //   if (CSS_STYLE.length > 0) newStyleCode += CSS_STYLE + '\n'
      //   if (INLINE_STYLE.length > 0) newStyleCode += INLINE_STYLE + '\n'
      //   if (UNO_STYLE.length > 0) newStyleCode += UNO_STYLE + '\n'
      let newCode = ''
      if (preflightStyles.length > 0) newCode += '\n' + preflightStyles
      if (safelistStyles.length > 0) newCode += '\n' + safelistStyles
      if (defaultStyles.length > 0) newCode += '\n' + defaultStyles
      if (attributifyStyles.length > 0) newCode += '\n' + attributifyStyles
      if (iconStyles.length > 0) newCode += '\n' + iconStyles
      if (cssStyles.length > 0) newCode += '\n' + cssStyles
      newCode += '\n'

      return {
        code: newCode,
      }
    },
  }
}

export interface BaseConfig {
  silent?: boolean
  mode?: 'development' | 'production'
  configPath?: string
  disableFormat?: boolean
  devTools?: {
    enabled: boolean
    completions?: boolean
  }
  //
  safeList?: string
  preflights?: boolean
  // bundle?: string;
  // debug?: boolean;
  // compile?: boolean;
  // prefix?: string;
  // verbosity?: number;
  experimental?: {
    icons?: {
      prefix?: string
      collections?: Record<string, IconifyJSON>
      extraProperties?: Record<string, string>
    }
    scan?: boolean
  }
}
export type DefaultConfig = BaseConfig
export type UserConfig = BaseConfig

const defaultConfig: DefaultConfig = {
  silent: false,
  // mode
  // configPath
  disableFormat: false,
  devTools: {
    enabled: false,
  },
  safeList: undefined,
  preflights: true,
  // bundle: undefined,
  // compile: false,
  // prefix: 'windi-',
  // verbosity: 1,
  // debug: false,
}

let configuration: BaseConfig
let generatorWindi: Processor
let generatorUno: UnoGenerator

// if (windiConfig != undefined) {
//   if (OPTIONS.configPath) {
//     // get modified time of config file
//     const mTime = statSync(OPTIONS.configPath).mtimeMs
//     if (mTime > configMTime) {
//       const tmpConfigPath = `./${Date.now()}windi.config.js`
//       copyFileSync(OPTIONS.configPath, tmpConfigPath)
//       loadConfig(tmpConfigPath)
//         .catch(e => {
//           useDebugger.createLog('Unknown Error while loading the config')
//           console.error(e)
//         })
//         .finally(() => {
//           rmSync(tmpConfigPath)
//           resolve({
//             code: _preprocess(content, filename),
//           })
//         })
//     } else {
//       resolve({
//         code: _preprocess(content, filename),
//       })
//     }
//   }
// } else {
//   if (OPTIONS.configPath) {
//     loadConfig(OPTIONS.configPath)
//       .catch(e => {
//         useDebugger.createLog('Unknown Error while loading the config')
//         console.error(e)
//       })
//       .finally(() => {
//         resolve({
//           code: _preprocess(content, filename),
//         })
//       })
//   } else {
//     PROCESSOR.loadConfig()
//     resolve({
//       code: _preprocess(content, filename),
//     })
//   }
// }

export function windi(userConfig: UserConfig = {}): PreprocessorGroup {
  configuration = { ...defaultConfig, ...userConfig }
  generatorWindi = new Processor()
  const steps: PreprocessorGroup[] = []
  if (configuration.experimental?.icons != undefined) {
    generatorUno = createGenerator(
      {
        presets: [
          UnocssIcons({
            ...configuration.experimental.icons,
          }),
        ],
      },
      {}
    )
  }

  if (configuration.experimental?.scan == undefined) {
    steps.push(agent())
  }
  steps.push(main())

  return {
    async markup({ content, filename }): Promise<Processed> {
      let code = content

      for (const step of steps) {
        code = (await preprocess(content, step, { filename })).code
      }

      return {
        code,
      }
    },
  }
}

export default windi
