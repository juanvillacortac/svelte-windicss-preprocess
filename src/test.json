[
  {
    "start": 0,
    "end": 103,
    "name": "html",
    "value": [
      {
        "start": 6,
        "end": 66,
        "type": "Attribute",
        "name": "class",
        "value": [
          {
            "start": 16,
            "end": 25,
            "data": "font-bold",
            "type": "Text"
          },
          {
            "start": 29,
            "end": 32,
            "data": "123",
            "type": "Expression"
          },
          {
            "start": 36,
            "end": 44,
            "data": "bg-white",
            "type": "Text"
          },
          {
            "start": 49,
            "end": 55,
            "data": "dafsdf",
            "type": "Expression"
          },
          {
            "start": 59,
            "end": 63,
            "data": "3232",
            "type": "Text"
          }
        ]
      },
      {
        "start": 68,
        "end": 102,
        "type": "Attribute",
        "name": "class",
        "value": [
          {
            "start": 75,
            "end": 101,
            "data": "bg-red-400 text-yellow-300",
            "type": "Text"
          }
        ]
      }
    ]
  },
  {
    "start": 103,
    "end": 212,
    "name": "p",
    "value": [
      {
        "start": 106,
        "end": 131,
        "type": "Attribute",
        "name": "class",
        "value": [
          {
            "start": 113,
            "end": 130,
            "data": "font-bold text-lg",
            "type": "Text"
          }
        ]
      },
      {
        "start": 132,
        "end": 155,
        "type": "Attribute",
        "name": "tw",
        "value": [
          {
            "start": 136,
            "end": 154,
            "data": "font-lg text-green",
            "type": "Text"
          }
        ]
      },
      {
        "start": 156,
        "end": 168,
        "type": "Attribute",
        "name": "class",
        "value": {
          "start": 163,
          "end": 167,
          "data": "helo",
          "type": "Expression"
        }
      },
      {
        "start": 169,
        "end": 194,
        "type": "Directive",
        "name": "bg-green-300",
        "value": {
          "start": 189,
          "end": 193,
          "data": "true",
          "type": "Expression"
        }
      },
      {
        "start": 195,
        "end": 211,
        "type": "Attribute",
        "name": "class",
        "value": {
          "start": 201,
          "end": 211,
          "data": "bg-red-500",
          "type": "Text"
        }
      }
    ]
  }
]