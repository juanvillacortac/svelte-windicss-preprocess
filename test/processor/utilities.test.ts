import { writeFileSync } from 'fs';
import { preprocess } from '../../src/index';

describe("utilities style test", () => {
  let result: string;
  beforeEach(async function () {
    const content = `<div class="min-h-screen bg-gray-100 py-6 justify-center aspect-w-9 aspect-h-16" tw="flex flex-col" sm="py-12"></div>`;
    result = (await preprocess().markup({ content, filename: 'test.svelte' })).code;
  });
  it("should generate utilities", () => {
    writeFileSync('test.css', result);
    expect(result).toMatchSnapshot('utilities');
  });
});