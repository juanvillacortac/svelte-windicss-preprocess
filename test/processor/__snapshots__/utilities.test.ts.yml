comment style test / should generate utilities / utilities / 0: >-
  <div   class="min-h-screen bg-gray-100 py-6 justify-center aspect-w-9
  aspect-h-16 flex flex-col sm:py-12"></div>


  <style>

  :global(*, ::before, ::after) {
    -webkit-box-sizing: border-box;
    box-sizing: border-box;
    border-width: 0;
    border-style: solid;
    border-color: #e5e7eb;
  }

  :global(*) {
    --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgba(59, 130, 246, 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
  }

  :global(:root) {
    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;
  }

  :global(:-moz-focusring) {
    outline: 1px dotted ButtonText;
  }

  :global(:-moz-ui-invalid) {
    box-shadow: none;
  }

  :global(::moz-focus-inner) {
    border-style: none;
    padding: 0;
  }

  :global(::-webkit-inner-spin-button, ::-webkit-outer-spin-button) {
    height: auto;
  }

  :global(::-webkit-search-decoration) {
    -webkit-appearance: none;
  }

  :global(::-webkit-file-upload-button) {
    -webkit-appearance: button;
    font: inherit;
  }

  :global([type='search']) {
    -webkit-appearance: textfield;
    outline-offset: -2px;
  }

  :global(abbr[title]) {
    -webkit-text-decoration: underline dotted;
    text-decoration: underline dotted;
  }

  :global(body) {
    margin: 0;
    font-family: inherit;
    line-height: inherit;
  }

  :global(html) {
    -webkit-text-size-adjust: 100%;
    font-family: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
    line-height: 1.5;
  }

  :global(.bg-gray-100) {
    --tw-bg-opacity: 1;
    background-color: rgba(243, 244, 246, var(--tw-bg-opacity));
  }

  :global(.flex) {
    display: -webkit-box;
    display: -ms-flexbox;
    display: -webkit-flex;
    display: flex;
  }

  :global(.flex-col) {
    -webkit-box-orient: vertical;
    -webkit-box-direction: normal;
    -ms-flex-direction: column;
    -webkit-flex-direction: column;
    flex-direction: column;
  }

  :global(.min-h-screen) {
    min-height: 100vh;
  }

  :global(.py-6) {
    padding-top: 1.5rem;
    padding-bottom: 1.5rem;
  }

  :global(.justify-center) {
    -webkit-box-pack: center;
    -ms-flex-pack: center;
    -webkit-justify-content: center;
    justify-content: center;
  }

  @media (min-width: 640px) {
    :global(.sm\:py-12) {
      padding-top: 3rem;
      padding-bottom: 3rem;
    }
  }

  </style>
